machine/v1beta1/types_azureprovider.go

```diff
@@ -83,6 +86,11 @@ type AzureMachineProviderSpec struct {
 	// SecurityProfile specifies the Security profile settings for a virtual machine.
 	// +optional
 	SecurityProfile *SecurityProfile `json:"securityProfile,omitempty"`
+	// UltraSSDCapability enables or disables Azure UltraSSD capability for a virtual machine.
+	// When omitted, the platform may enable the capability based on the configuration of data disks.
+	// +kubebuilder:validation:Enum:="Enabled";"Disabled"
+	// +optional
+	UltraSSDCapability AzureUltraSSDCapabilityState `json:"ultraSSDCapability,omitempty"`
```

Comment:
Booleans are not allowed based on the OpenShift and Kubernetes API conventions. In particular, having booleans does not allow for a third state to determine if the user has not set the value. For example, here, we will imply from the fact that they have omitted the value but requested an UltraDisk for the OSDisk that it should be enabled.

In general, we aren't restricted by the upstream APIs of the cloud provider and should aim to provide something that is clear to users, I'm pretty happy that Enabled and Disabled are pretty obvious to end users

Reply:
What I said is not excluded by the k8s convention... as `UltraSSDEnabled` can be `{'true','false',''}` and would be an obvious mapping to the targeted API parameter.

Reply:
Would having a string value in the schema with `true`, `false` and `''` values work? I would have expected marshalling issues with that kind of schema? Something along the lines of `expected string, got bool`? Maybe it does work

Reply:
Coming back to this, I'm still a firm believer that we should try to make our APIs readable and clear, and while having this different from the upstream Azure API is a source of possible confusion, I think the naming here is clear enough and, personally, more in-keeping with kube practices and naming. I don't believe this will be a source of confusion for users, and I think having the name this way makes it expandable in the future should the capability change at all

Reply:
I agree with Joel here, I'd rather not have booleans encoded as strings (can be misleading for the user and a source of error) just for the sake of keeping the same name present in the Azure API.

Reply:
I was only trying to help by providing my insight based on my 20+ years of experience. As I have two cards blocked on this PR, I'll gladly step aside in favor of getting this merged sooner than later.

Reply:
@jstuever We definitely appreciate your input here and thank you for raising the topic for discussion. 

My team have been working very closely with the API review team over the last few months and in this particular case, are following advice from previous discussions about the inflexibility of boolean fields and how we don't have to be restricted by upstream API decisions, and should consider the UX for users of our systems to make our configuration APIs easy to understand, and in keeping with our conventions.

We discussed as a team the term `Enabled` and how that wouldn't fit in with Kubernetes naming conventions and how with that term in the name, it would mean the only possible values really are true and false. By moving the state of the capability into the value, we allow for future extensibility of the API field should new context arise (eg we may want an autodetect value, which is the current behaviour when omitted), and follow the patterns set out in prior APIs within OpenShift. 

As we are already close to merging this PR and have gone through several rounds of API review already, I'm going to ask @damdo to resolve this conversation, but if you have any further concerns, I'm happy to have a chat.